tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -0,0 +1,322 @@
+    do {
+      t_evaluations.push_back(cur);
+      cur *= step;
+    } while (cur != orig);
+    CHECK_EQ(t_evaluations.size(),
+             size_t{1} << (extended_domain->log_size_of_group() -
+                           domain->log_size_of_group()));
+
+    // Subtract 1 from each to give us t_evaluations[i] = t(zeta *
+    // extended_omegaⁱ)
+    base::Parallelize(t_evaluations, [](absl::Span<F> chunk) {
+      for (F& coeff : chunk) {
+        coeff -= F::One();
+      }
+    });
+
+    F::BatchInverseInPlace(t_evaluations);
+
+    // Multiply the inverse to obtain the quotient polynomial in the coset
+    // evaluation domain.
+    // TODO(TomTaehoonKim): Avoid copy here by adding element wise
+    // multiplication.
+    std::vector<F> evaluations = evals.evaluations();
+    base::Parallelize(evaluations,
+                      [t_evaluations](absl::Span<F> chunk, size_t chunk_idx,
+                                      size_t chunk_size) {
+                        size_t index = chunk_idx * chunk_size;
+                        for (F& h : chunk) {
+                          h *= t_evaluations[index % t_evaluations.size()];
+                          ++index;
+                        }
+                      });
+
+    return ExtendedEvals(std::move(evaluations));
+  }
+
+  // This takes us from the extended evaluation domain and gets us the quotient
+  // polynomial coefficients.
+  //
+  // This function will panic if the provided vector is not the correct length.
+  constexpr ExtendedPoly ExtendedToCoeff(
+      const ExtendedEvals& evals, const ExtendedDomain* extended_domain) const {
+    CHECK_EQ(evals.NumElements(), extended_domain->size());
+
+    ExtendedPoly poly = extended_domain->IFFT(evals);
+
+    // Distribute powers to move from coset; opposite from the
+    // transformation we performed earlier.
+    DistributePowersZeta(poly, false);
+
+    return poly;
+  }
+
+  // Given a |poly| of coefficients  [a₀, a₁, a₂, ...], this returns
+  // [a₀, ζa₁, ζ²a₂, a₃, ζa₄, ζ²a₅, a₆, ...], where ζ is a cube root of unity in
+  // the multiplicative subgroup with order (p - 1), i.e. ζ³ = 1.
+  //
+  // |into_coset| should be set to true when moving into the coset, and false
+  // when moving out. This toggles the choice of ζ.
+  void DistributePowersZeta(ExtendedPoly& poly, bool into_coset) const {
+    // Calculate ζ = g^((2ˢ * T) / 3)
+    CHECK_EQ(F::Config::kTrace % math::BigInt<F::kLimbNums>(3),
+             math::BigInt<F::kLimbNums>(0));
+    F zeta_inv = F::FromMontgomery(F::Config::kSubgroupGenerator)
+                     .Pow(F(2).Pow(F::Config::kTwoAdicity).ToBigInt() *
+                          F::Config::kTrace / math::BigInt<F::kLimbNums>(3));
+    F zeta = zeta_inv.Square();
+    std::vector<F> coset_powers{into_coset ? zeta : zeta_inv,
+                                into_coset ? zeta_inv : zeta};
+
+    // FIXME(TomTaehoonKim): Avoid copy.
+    std::vector<F> coeffs = poly.coefficients().coefficients();
```

Comment:
Are you worried about copying here?

Reply:
Yes, but I couldn't find a solution since there's no way I can make changes directly on the coefficients.